graphql/src/main/java/com/linecorp/armeria/server/graphql/DefaultGraphQLService.java

```diff
@@ -0,0 +1,204 @@
+                }
+                final String query = (String) inputMap.getOrDefault("query", "");
+                final Map<String, Object> variables =
+                        (Map<String, Object>) inputMap.getOrDefault("variables", ImmutableMap.of());
+                final Map<String, Object> extensions =
+                        (Map<String, Object>) inputMap.getOrDefault("extensions", ImmutableMap.of());
+                return execute(ctx, GraphQLInput.builder(query, ctx)
+                                                .variables(variables)
+                                                .operationName((String) inputMap.get("operationName"))
+                                                .extensions(extensions)
+                                                .dataLoaderRegistry(dataLoaderRegistry)
+                                                .build());
+            }));
+        } else if (contentType != null &&
+                   contentType.type().equals(APPLICATION_GRAPHQL.type()) &&
+                   contentType.subtype().equals(APPLICATION_GRAPHQL.subtype())) {
+            return HttpResponse.from(request.aggregate().handleAsync((req, thrown) -> {
+                if (thrown != null) {
+                    logger.warn("{} Failed to aggregate a request:", ctx, thrown);
+                    return HttpResponse.ofFailure(thrown);
+                }
+                final String query = req.contentUtf8();
+                return execute(ctx, GraphQLInput.builder(query, ctx)
+                                                .dataLoaderRegistry(dataLoaderRegistry)
+                                                .build());
+            }));
+        } else {
+            return HttpResponse.of(HttpStatus.UNPROCESSABLE_ENTITY,
+                                   MediaType.PLAIN_TEXT,
+                                   "Could not process GraphQL request");
+        }
+    }
+
+    private HttpResponse execute(ServiceRequestContext ctx, GraphQLInput input) {
+        try {
+            final GraphQLOutput output = executor.serve(ctx, input);
+            return HttpResponse.of(MediaType.JSON_UTF_8,
+                                   deserialize(output.toSpecification()));
+        } catch (Exception e) {
+            final ExecutionResultImpl executionResult = new ExecutionResultImpl(
+                    GraphqlErrorException.newErrorException()
+                                         .message(e.getMessage())
+                                         .cause(e)
+                                         .build());
+            final GraphQLOutput output = GraphQLOutput.of(executionResult);
+            return HttpResponse.of(MediaType.JSON_UTF_8, deserialize(output.toSpecification()));
+        }
+    }
+
+    @Override
+    public Set<Route> routes() {
+        return ImmutableSet.of(route);
+    }
+
+    private static Map<String, Object> serialize(@Nullable String content) {
```

Comment:
FYI. Technically, this is `deserialize`. 😉
https://en.wikipedia.org/wiki/Serialization

How about giving a more specific name such as `parseJsonString()`?

Reply:
Oops... 🤭